### PR TITLE
Several improvements related to  Eigen Tensors

### LINKED ISSUE #1013: Segmentation fault when accessing Essentia tensors from Python
Right now our `tensorReal` type is designed to serve Numpy array references (without coping the data) to Python.
https://github.com/MTG/essentia/blob/ba79be6515f2fd0cde75ee3f6fa98706a66f4c36/src/python/pytypes/tensorreal.cpp#L29

However, this scenario never happens as our only algorithm returning tensors (`PoolToTensor`) is only defined in streaming mode. Thus, in Python these tensors  have to be interfaced with `Pools`, where the tensor's data is copied before making the Numpy reference.
https://github.com/MTG/essentia/blob/ba79be6515f2fd0cde75ee3f6fa98706a66f4c36/src/python/pytypes/pypool.cpp#L534

Working on a standard algorithm returning a tensor, I have found memory deallocation errors when trying to access tensors' references in Python.

Given that we are already coping the tensors' data in the `Pools`, a solution would be to generalize this approach by changing `tensorReal`'s `toPythonRef` method to a `toPythonCopy` one.

### PR DESCRIPTION
**Pull request by pabloEntropia**
This PR contains several improvements related to  Eigen Tensors:

- TensorReal toPythonRef -> toPythonCopy to fix #1013 
- Add convenience aliases for tensors of different ranks
- Disable build-time Eigen-related warnings
- Add basic statistic operations with tensors (mean, std, min, max)
- Add new Algorithms `TensorTranspose` & `TensorNormalize`

### CODE REVIEW

File: src/essentia/essentiamath.h [hunk 1]
```diff
@@ -1200,6 +1200,142 @@ std::vector<std::vector<T> > pairwiseDistance(const std::vector<std::vector<T> >
...
 }
 
+/**
+ * Sets `squeezeShape`, `summerizerShape`, `broadcastShape` to perform operations 
+ * on a Tensor with the shape of `tesor` along the `axis` dimension.
+ */
+template <typename T>
+void tensorGeometricalInfo(const Tensor<T>& tensor, int& axis,
+                           std::array<Eigen::Index, TENSORRANK - 1>& squeezeShape,
+                           std::array<Eigen::Index, TENSORRANK>& summerizerShape,
```
Comment (dbogdanov) [inline src/essentia/essentiamath.h:1210]: "summarizerShare" (typo)

File: src/essentia/essentiamath.h [hunk 2]
```diff
@@ -1200,6 +1200,142 @@ std::vector<std::vector<T> > pairwiseDistance(const std::vector<std::vector<T> >
   return pdist;
 }
 
+/**
+ * Sets `squeezeShape`, `summerizerShape`, `broadcastShape` to perform operations 
+ * on a Tensor with the shape of `tesor` along the `axis` dimension.
+ */
+template <typename T>
+void tensorGeometricalInfo(const Tensor<T>& tensor, int& axis,
```
Comment (dbogdanov) [inline src/essentia/essentiamath.h:1208]: Do we need to group computation for `squeezeShape`, `summarizerShape`and `broadcastShape` together in one method?

I guess it is fine if we use this mostly as an auxiliary method.

File: src/essentia/essentiamath.h [hunk 3]
```diff
@@ -1200,6 +1200,142 @@ std::vector<std::vector<T> > pairwiseDistance(const std::vector<std::vector<T> >
   return pdist;
 }
 
+/**
+ * Sets `squeezeShape`, `summerizerShape`, `broadcastShape` to perform operations 
+ * on a Tensor with the shape of `tesor` along the `axis` dimension.
```
Comment (dbogdanov) [inline src/essentia/essentiamath.h:1205]: typo: `tesor` -> `tensor`

File: src/essentia/essentiamath.h (hunk 2 above)
Comment (dbogdanov) [inline src/essentia/essentiamath.h:1208]: We can consider simplify the name: `Geometrical` -> `Geometric`. Also for consistency with other names (e.g., we have the `GeometricMean` algorithm)

File: test/src/unittests/standard/test_tensornormalize.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+class TestTensorNormalize(TestCase):
+
+    def scalerSingleValue(self, scaler_arg=('standard', StandardScaler())):
+        scaler_name, scaler = scaler_arg
+
+        original = numpy.arange(1, dtype='float32')
+
+        # Use Scipy to generate the expected results
+        expected =scaler.fit_transform(original.reshape(-1, 1))
```
Comment (dbogdanov) [inline test/src/unittests/standard/test_tensornormalize.py:34]: " = "

File: test/src/unittests/standard/test_tensornormalize.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+# You should have received a copy of the Affero GNU General Public License
+# version 3 along with this program. If not, see http://www.gnu.org/licenses/
+
+
+
+from essentia_test import *
+from sklearn.preprocessing import MinMaxScaler, StandardScaler
+
+
+class TestTensorNormalize(TestCase):
```
Comment (dbogdanov) [inline test/src/unittests/standard/test_tensornormalize.py:26]: What happens in the extreme cases (e.g., all value are the same)?

File: src/algorithms/standard/tensortranspose.cpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+const char* TensorTranspose::name = "TensorTranspose";
+const char* TensorTranspose::category = "Standard";
+const char* TensorTranspose::description = DOC("This algorithm performs transpositions over the axes of a tensor.\n");
+
+  void TensorTranspose::configure() {
+    if (!parameter("permutation").isConfigured()) return;
+    _permutation = parameter("permutation").toVectorInt();
+
+    if (_permutation.size() != TENSORRANK) {
```
Comment (dbogdanov) [inline src/algorithms/standard/tensortranspose.cpp:36]: These checks do not take into consideration the case of  having the same value repeated in the permutation vector.

File: src/algorithms/standard/tensornormalize.h [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  int _axis;
+
+ public:
+  TensorNormalize() {
+    declareInput(_input, "tensor", "the input tensor");
+    declareOutput(_output, "tensor", "the normalized output tensor");
+  }
+
+  void declareParameters() {
+    declareParameter("scaler", "the type of the normalization to apply to input tensor", "{standard,minMax}", "standard");
```
[the remark this answers is not in the document]
Comment (pabloEntropia) [reply src/algorithms/standard/tensornormalize.h:54]: I borrowed the names from scikit-learn methods:
 - [StandardScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html)
- [MinMaxScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html)

Also, in the algorithm documentation I refer to [this](https://en.wikipedia.org/wiki/Feature_scaling) Wikipedia article that calls it `Standardization`.

About the parameter values, I have found 14 parameters using values with underscore and 9 using lower camel case.
Interestingly, some algorithms like `MelBands` and `MFCC` have parameters with both styles.

{slaneyMel,htkMel} -> MelBands, MFCC
{none,unitSum,unitMax} -> HPCP
{none,cosine,squaredCosine} -> KeyExtractor, HPCP
{linear,slaneyMel,htkMel} -> TriangularBands
{sumEnergy, peakEnergy} -> SingleBeatLoudness

{none,unit_sum,unit_max} -> Chromagram, Histogram
{starting_beat,interbeat_median} -> ChordDetectionBeats
{unit_sum,unit_tri,unit_max} -> MelBands, MFCC, TriangularBands
{unit_sum,unit_max} -> BFCC, SpectrumToCent, TriangularBarkBands
{flat,triangle,inverse_triangle,parabola,inverse_parabola,linear,quadratic,inverse_quadratic,hybrid,supplied} -> NoveltyCurve
{infogain,beat_emphasis} -> OnsetDetectionGlobal
{hfc,complex,complex_phase,flux,melflux,rms} -> OnsetDetection

We should decice something about this. I would prefer lower camel case for consistency with the parameter names, although underscores seem more pythonic.

File: src/essentia/essentiamath.h (hunk 2 above)
Comment (pabloEntropia) [reply src/essentia/essentiamath.h:1208]: Yes, I thought it could be a useful auxiliary method in case we need to implement more axis-wise tensor operations in the future.

Comment (pabloEntropia) [reply src/essentia/essentiamath.h:1208]: I used `geometrical`  for consistency with  Eigen [Geometrical operations ](https://eigen.tuxfamily.org/dox/unsupported/eigen_tensors.html#autotoc_md95), but I do not have a preference for one name or another.

File: src/algorithms/standard/tensornormalize.cpp [hunk 8]
```diff
@@ -56,15 +56,35 @@ void TensorNormalize::compute() {
...
       if (_axis == -1) {
-        Real means = mean(input);
-        Real stds = stddev(input, means);
+        Real globalMean = mean(input);
+        Real globalStd = stddev(input, globalMean);
 
-        output = (input - input.constant(means)) / input.constant(stds);
+        if (globalStd == 0) {
+          E_INFO("TensorNormalize: Received tensor with constant value.");
+          // Set std to 0 so we return a vector of all 0s.
```
Comment (dbogdanov) [inline src/algorithms/standard/tensornormalize.cpp:64]: Typo here? (should be "Set std to 1...")?

File: src/essentia/essentiamath.h (hunk 2 above)
Comment (dbogdanov) [reply src/essentia/essentiamath.h:1208]: Then, this is fine
